physics/kepler_orbit_body.hpp

```diff
@@ -22,27 +23,95 @@ KeplerOrbit<Frame>::KeplerOrbit(
+// file.
 
 template<typename Frame>
 RelativeDegreesOfFreedom<Frame>
 KeplerOrbit<Frame>::StateVectors(Instant const& t) const {
-  GravitationalParameter const μ = system_gravitational_parameter_;
-  double const eccentricity = elements_at_epoch_.eccentricity;
-  Length const a = elements_at_epoch_.semimajor_axis;
-  Angle const i = elements_at_epoch_.inclination;
-  Angle const Ω = elements_at_epoch_.longitude_of_ascending_node;
-  Angle const ω = elements_at_epoch_.argument_of_periapsis;
-  AngularFrequency const mean_motion = Sqrt(μ / Pow<3>(a)) * Radian;
-  Angle const mean_anomaly = elements_at_epoch_.mean_anomaly +
-                             mean_motion * (t - epoch_);
+  KeplerianElements<Frame> primocentric_elements = elements_at_epoch_;
+
+  GravitationalParameter const μ = primary_gravitational_parameter_ +
+                                   secondary_gravitational_parameter_;
+
+  // Update the mean anomaly for |t|.
+  primocentric_elements.mean_anomaly = elements_at_epoch_.mean_anomaly +
+                                         mean_motion() * (t - epoch_);
+
+  return TestParticleStateVectors(primocentric_elements, μ);
+}
+
+template<typename Frame>
+RelativeDegreesOfFreedom<Frame>
+KeplerOrbit<Frame>::BarycentricStateVectors(Instant const& t) const {
+  KeplerianElements<Frame> barycentric_elements = elements_at_epoch_;
+
+  // μ is such that the mean motion (and thus the period) is the same as for the
+  // primocentric orbit, μ/a^3 = (μ1 + μ2)/(a_primocentric)^3, where |a| is the
+  // semimajor axis of the orbit around the barycentre.
+  GravitationalParameter const μ1 = primary_gravitational_parameter_;
+  GravitationalParameter const μ2 = secondary_gravitational_parameter_;
+  GravitationalParameter const μ = Pow<3>(μ1) / Pow<2>(μ1 + μ2);
+
+  if (barycentric_elements.conic.semimajor_axis) {
+    // Change the semimajor axis to get elements describing the orbit of the
+    // secondary around the barycentre, rather than around the primary.
+    Length const a_primocentric = *barycentric_elements.conic.semimajor_axis;
+    barycentric_elements.conic.semimajor_axis = a_primocentric * μ1 / (μ1 + μ2);
+  }
+
+  barycentric_elements.mean_anomaly = elements_at_epoch_.mean_anomaly +
+                                      mean_motion() * (t - epoch_);
+
+  return TestParticleStateVectors(barycentric_elements, μ);
+}
+
+template<typename Frame>
+AngularFrequency KeplerOrbit<Frame>::mean_motion() const {
+  GravitationalParameter const μ = primary_gravitational_parameter_ +
+                                   secondary_gravitational_parameter_;
+  if (elements_at_epoch_.conic.mean_motion) {
+    return *elements_at_epoch_.conic.mean_motion;
+  } else {
+    GravitationalParameter const μ = primary_gravitational_parameter_ +
+                                     secondary_gravitational_parameter_;
+    Length const a = *elements_at_epoch_.conic.semimajor_axis;
+    return Sqrt(μ / Pow<3>(a)) * Radian;
```

Comment:
Put this in a member function of Conic: give-me-the-mean-motion.  If Conic is here to stay, might as well give it a bit more abstraction.